modules/assetfs/layered.go

```diff
@@ -0,0 +1,266 @@
+// The first layer is the top layer, and it will be used first.
+// If the file is not found in the top layer, it will be searched in the next layer.
+type LayeredFS struct {
+	layers []*Layer
+}
+
+// Layered returns a new LayeredFS with the given layers. The first layer is the top layer.
+func Layered(layers ...*Layer) *LayeredFS {
+	return &LayeredFS{layers: layers}
+}
+
+// Open opens the named file. The caller is responsible for closing the file.
+func (l *LayeredFS) Open(name string) (http.File, error) {
+	for _, layer := range l.layers {
+		f, err := layer.Open(name)
+		if err == nil || !os.IsNotExist(err) {
+			return f, err
+		}
+	}
+	return nil, fs.ErrNotExist
+}
+
+// ReadFile reads the named file.
+func (l *LayeredFS) ReadFile(elems ...string) ([]byte, error) {
+	bs, _, err := l.ReadLayeredFile(elems...)
+	return bs, err
+}
+
+// ReadLayeredFile reads the named file, and returns the layer name.
+func (l *LayeredFS) ReadLayeredFile(elems ...string) ([]byte, string, error) {
+	name := util.PathJoinRel(elems...)
+	for _, layer := range l.layers {
+		f, err := layer.Open(name)
+		if os.IsNotExist(err) {
+			continue
+		} else if err != nil {
+			return nil, layer.name, err
+		}
+		bs, err := io.ReadAll(f)
+		_ = f.Close()
+		return bs, layer.name, err
+	}
+	return nil, "", fs.ErrNotExist
+}
+
+// ListFiles lists files/directories in the given directory. The fileMode controls the returned files.
+// * omitted: all files and directories will be returned.
+// * true: only files will be returned.
+// * false: only directories will be returned.
+func (l *LayeredFS) ListFiles(name string, fileMode ...bool) ([]string, error) {
+	fileMap := map[string]bool{}
+	for _, layer := range l.layers {
+		f, err := layer.Open(name)
+		if os.IsNotExist(err) {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		infos, err := f.Readdir(-1)
+		_ = f.Close()
+		if err != nil {
+			return nil, err
+		}
+		for _, info := range infos {
+			include := false
+			if len(fileMode) == 0 {
+				include = true
+			} else if len(fileMode) == 1 && fileMode[0] == !info.Mode().IsDir() {
+				include = true
+			} else if len(fileMode) > 1 {
+				return nil, errors.New("too many arguments")
+			}
+			if include {
+				fileMap[info.Name()] = true
+			}
+		}
+	}
+	var files []string
+	for file := range fileMap {
+		files = append(files, file)
+	}
+	sort.Strings(files)
+	return files, nil
+}
+
+// ListAllFiles returns files/directories  in the given directory, including subdirectories, recursively.
+// The fileMode controls the returned files:
+// * omitted: all files and directories will be returned.
+// * true: only files will be returned.
+// * false: only directories will be returned.
+func (l *LayeredFS) ListAllFiles(name string, fileMode ...bool) ([]string, error) {
+	return l.listAllFiles(l.layers, name, fileMode...)
+}
+
+func (l *LayeredFS) listAllFiles(layers []*Layer, name string, fileMode ...bool) ([]string, error) {
```

Comment:
```suggestion
func listAllFiles(layers []*Layer, name string, fileMode ...bool) ([]string, error) {
```

Reply:
done in 96619fdd1c0946a0058fa321fdda77e22ca8eb9b